kore/src/Kore/Unification/NewUnifier.hs

```diff
@@ -255,10 +258,25 @@ substituteInKeys ::
 substituteInKeys x s solution =
     foldr substitute' ([], []) solution
   where
-    substitute' subst (eqs, solution') =
+    substitute' subst ~(eqs, solution') =
+        -- We seem to need an explicit inline to achieve specialization of alterF.
+        -- That sounds like a containers problem.
+        case inline Map.alterF (\case Nothing -> NotFound; Just t -> Found t Nothing) x subst of
+            NotFound -> (eqs, subst : solution')
+            Found t !subst' -> ((s, t) : eqs, subst' : solution')
+
+{-
+-- The above just does the following, but without needing to compare variable
+-- names in the second (deletion) pass.
+
         case Map.lookup x subst of
             Nothing -> (eqs, subst : solution')
-            Just t -> ((s, t) : eqs, Map.delete x subst : solution')
+            Just t ->
+              let !subst' = Map.delete x subst
+              in ((s, t) : eqs, subst' : solution')
+              -}
+data CheckRes a m = NotFound | Found !a ~m
+    deriving stock (Functor)
```

Comment:
Is there a utilities module I can stick a function to look up a key in a map and delete it if it's found? This module doesn't really seem quite the place for it.